lgc/patch/PatchBufferOp.cpp

```diff
@@ -1617,21 +1682,38 @@ Value *PatchBufferOp::replaceLoadStore(Instruction &inst) {
     }
   }
 
+  const DataLayout &DL = m_builder->GetInsertBlock()->getModule()->getDataLayout();
+  bool canUseSBufferLoadImm = m_pipelineState->getTargetInfo().getGfxIpVersion().major >= 9;
+
   // The index in storeValue which we use next
   unsigned storeIndex = 0;
 
   unsigned remainingBytes = bytesToHandle;
   while (remainingBytes > 0) {
-    const unsigned offset = bytesToHandle - remainingBytes;
-    Value *offsetVal = offset == 0 ? baseIndex : m_builder->CreateAdd(baseIndex, m_builder->getInt32(offset));
 
     Type *intAccessType = getIntAccessType(m_context, alignment, isInvariant, remainingBytes);
-    const DataLayout &DL = m_builder->GetInsertBlock()->getModule()->getDataLayout();
     unsigned accessSize = DL.getTypeSizeInBits(intAccessType) / 8;
+    unsigned intrinsicID = getLSIntrinsicID(isLoad, isInvariant, accessSize, ordering);
 
-    Value *part = nullptr;
+    auto offsetPair = std::make_pair(baseIndex, bytesToHandle - remainingBytes);
 
-    unsigned intrinsicID = getLSIntrinsicID(isLoad, isInvariant, accessSize, ordering);
+    // FIXME: in addition to SMEM loads, we could utilize immediate offset field in VMEM loads.
+
+    // Try using amdgcn_s_buffer_load_imm instead of amdgcn_s_buffer_load if target supports it and
+    // the immediate part of the offset is non-zero.
+    if (canUseSBufferLoadImm && intrinsicID == Intrinsic::amdgcn_s_buffer_load) {
```

Comment:
Please add a note here that we should try to remove this again once we have GlobalISel.

Reply:
Will do.